central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -523,7 +536,137 @@ func (ds *datastoreImpl) RemovePlopsByPod(ctx context.Context, id string) error
 	} else if !ok {
 		return sac.ErrResourceAccessDenied
 	}
+
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
 	q := search.NewQueryBuilder().AddExactMatches(search.PodUID, id).ProtoQuery()
 	_, storeErr := ds.storage.DeleteByQuery(ctx, q)
 	return storeErr
 }
+
+// PruneOrphanedPLOPs prunes old closed PLOPs and those without deployments or pods
+func (ds *datastoreImpl) PruneOrphanedPLOPs(ctx context.Context, orphanWindow time.Duration) int64 {
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	query := fmt.Sprintf(pruneOrphanedPLOPs, int(orphanWindow.Minutes()))
+	commandTag, err := ds.pool.Exec(ctx, query)
+	if err != nil {
+		log.Errorf("failed to prune PLOP: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned due to missing deployments
+	if _, err := ds.pool.Exec(ctx, deleteOrphanedPLOPDeployments); err != nil {
+		log.Errorf("failed to prune process listening on ports by deployment: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned due to missing pods.
+	if _, err := ds.pool.Exec(ctx, deleteOrphanedPLOPPodsWithPodUID); err != nil {
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+
+	return commandTag.RowsAffected()
+}
+
+// PruneOrphanedPLOPsByProcessIndicators prunes PLOPs that match process indicators without pods
+func (ds *datastoreImpl) PruneOrphanedPLOPsByProcessIndicators(ctx context.Context, orphanWindow time.Duration) {
+	// TODO(ROX-22443): Once it is guaranteed that all listening endpoints have PodUIDs, remove this function
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing deployments
+	query := fmt.Sprintf(deleteOrphanedPLOPDeploymentsAndPI, int(orphanWindow.Minutes()))
+	if _, err := ds.pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by deployment: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing pods.
+	query = fmt.Sprintf(deleteOrphanedPLOPPods, int(orphanWindow.Minutes()))
+	if _, err := ds.pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+}
+
+func (ds *datastoreImpl) readRowsToFindPLOPsWithNoProcessInformation(rows pgx.Rows) ([]string, error) {
+	var ids []string
+
+	for rows.Next() {
+		var serialized []byte
+
+		if err := rows.Scan(&serialized); err != nil {
+			return nil, pgutils.ErrNilIfNoRows(err)
+		}
+
+		var msg storage.ProcessListeningOnPortStorage
+		if err := proto.Unmarshal(serialized, &msg); err != nil {
+			return nil, err
+		}
+
+		process := msg.GetProcess()
+
+		if process == nil {
+			ids = append(ids, msg.GetId())
+		}
+	}
+	return ids, nil
+}
+
+func (ds *datastoreImpl) deletePLOPsInBatches(ctx context.Context, ids []string) (int64, error) {
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+	batchSize := 5000
```

Comment:
I don't think we need to do batching here.  `DeleteMany` appears to do batching on its own.

Reply:
I thought it might be better to do batching here, so that the lock isn't applied for too long, but is applied multiple times for shorter times instead. I think deletion is fast, so it should not be a problem just to use DeleteMany once.

Reply:
Right, but in order to that to work, you would have had to do the batching in the caller to this.  Shouldn't be a huge issue as this should not be called too frequently.